### DISCUSSION TITLE
3d Inflate effect

### DISCUSSION THREAD

**Discussion by alxpnv** [ImageMagick/ImageMagick#7037] [Digital Image Processing]
Hi there,

I'm trying to reproduce the Adobe Illustrator Inflate effect using Distance Morphology and `-shade` operation.
Here is what I've come up with so far:
```
convert heart.png -bordercolor transparent -border 20 heart_extended.png
convert heart_extended.png -alpha extract heart_alpha.png

convert heart_alpha.png -morphology Distance Euclidean:4,500 -fx 'sqrt(1 - (p - 1)^2)' heart_distance.png
# I've tried using the image from the previous step here as a source, but the result looks too flat
convert heart_alpha.png -blur 0x6 -shade 110x30 -normalize heart_distance.png -compose Overlay -composite heart_shaded.png
convert heart_shaded.png heart_extended.png -compose Multiply -composite heart_alpha.png -compose CopyOpacity -composite -shave 20 heart3d.png
```
![heart.png](https://github.com/ImageMagick/ImageMagick/assets/78347945/33c49f66-8761-4be2-9339-e38ae4eafd50)
![heart3d.png](https://github.com/ImageMagick/ImageMagick/assets/78347945/4805e621-7fa8-4fe2-a635-169267370fb7)

Do you have any suggestions to make the results more realistic (add proper highlights)?

**Comment by fmw42** [comment]
What is the expected result from Illustrator?  Can you show that image?

**Reply by alxpnv** [reply]
I don't have Illustrator, so here is a random image I've googled. Basically, Illustrator converts the object to a 3d mesh using distance from edges to determine height, then renders it using ray-tracing, so I don't expect a physically correct result, but something that looks more or less like an inflated shape. Actually, Photoshop CS5 Extended [had](https://abduzeedo.com/node/26798) this effect back in 2010, but it was removed later
![star](https://github.com/ImageMagick/ImageMagick/assets/78347945/2f7faa58-5f9e-45bd-afb2-2335974f323d)

**Comment by snibgo** [comment]
> Do you have any suggestions to make the results more realistic (add proper highlights)?

I suggest using a ray-tracer. See examples at [Coffee mug](http://im.snibgo.com/coffmug.htm).

**Comment by fmw42** [comment]
See examples at http://www.fmwconcepts.com/imagemagick/texteffect2/index.php especially the last set for the "chrome" examples.  Does any of that help you?

**Comment by fmw42** [comment]
See also "glass bubble button" at https://imagemagick.org/Usage/thumbnails/#glass_bubble and "badge lighting effects" at https://imagemagick.org/Usage/thumbnails/#badge_lighting

**Comment by alxpnv** [comment]
@fmw42 Thanks for the help. I'd like to get a more round look, but it seems that with `-shade` you can only imitate surfaces with linear cross-sections like cones, but not spheres

**Comment by snibgo** [comment]
> ... but it seems that with -shade you can only imitate surfaces with linear cross-sections like cones, but not spheres.

Not so. `-shade` is calculated from local gradients, and it doesn't matter how they have been created. For example, from this image `gr_pond.png`:

![gr_pond](https://github.com/ImageMagick/ImageMagick/assets/33812368/03fc6142-cd69-496b-9e6a-809aa288b015)

... we apply a shade:
```
magick gr_pond.png -shade 60x30 x.png
```
The result, `x.png`, is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/60efd3b6-865d-4d73-82d4-b8d441ed5266)

The main limitations of shade is that it uses a simple model of surface reflection, and only a single light source at infinity. Perhaps you could process the image to simulate specular reflection.

**Comment by fmw42** [comment]
You can adjust the result by controlling the function applied to the linear ramp (gradient) effect. For example:

Input

![red_heart](https://github.com/ImageMagick/ImageMagick/assets/13458218/1abc9d98-b6f2-49a9-a2a6-cd48b5cd6b07)


FX function:

```
convert red_heart.png \
\( +clone -alpha extract \
-gamma 2 +level 0,100 -white-threshold 99 \
-morphology Distance Euclidean:4 -auto-level \
-fx "sqrt(1-(u-1)*(u-1))" \
-shade 135x10 -auto-level +level 10,90% \) \
-compose hard_light -composite result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/e543f61b-de7d-41ee-944a-a16d9a33c0f1)


Sinusoid function:

```
convert red_heart.png \
\( +clone -alpha extract \
-gamma 2 +level 0,100 -white-threshold 99 \
-morphology Distance Euclidean:4 -auto-level \
-evaluate sin 1 \
-shade 135x10 -auto-level +level 10,90% \) \
-compose hard_light -composite result2.png
```

![result2](https://github.com/ImageMagick/ImageMagick/assets/13458218/6a2a5cf7-7484-4211-a5b9-7d64530b2110)


I leave it to you to adjust as desired

**Comment by fmw42** [comment]
Here is another way

```
rounding=100
convert red_heart.png \
\( +clone -alpha Extract -write mpr:alpha \
-blur 0x$rounding -shade 135x10 mpr:alpha \
-compose over -compose copy_opacity -composite \
-alpha deactivate -auto-level \
-function polynomial 3.5,-5.05,2.05,0.25 \
-alpha on \) \
-compose Hardlight -composite \
result2.png

```

![result2](https://github.com/ImageMagick/ImageMagick/assets/13458218/e887548a-a9fc-4317-853d-46d88a394969)

**Comment by fmw42** [comment]
Is the following what you want?

I have a bash unix shell script called bump.

Input:

![Beetle_Art](https://github.com/ImageMagick/ImageMagick/assets/13458218/6ebfa0b2-3d1d-4875-996e-c72ea4b74829)

`bump -t 3 -a 50 Beetle_Art.png Beetle_Art_bump.png
`

![Beetle_Art_bump](https://github.com/ImageMagick/ImageMagick/assets/13458218/51791161-f7cd-4d00-b86c-41fddb7429b1)

**Comment by alxpnv** [comment]
I ended up writing a [script](https://github.com/alxpnv/inflatesvg) which converts an SVG file to an inflated 3D mesh:
![heart](https://github.com/ImageMagick/ImageMagick/assets/78347945/16b7af0a-8ab4-4a99-aba0-95b85ebdd3cc)
